### PR TITLE
Feature/fct2 6951 implement stateless file transfer endpoint

### PR DESCRIPTION
**Pull request by NeilFoubister**

### CODE REVIEW

File: backend/CPS.ComplexCases.Egress.Tests/Integration/EgressStorageClientTests.cs [hunk 1]
```diff
@@ -233,7 +234,7 @@ public async Task InitiateUploadAsync_ShouldThrowArgumentNullException_WhenSourc
             () => _client.InitiateUploadAsync(destinationPath, fileSize, workspaceId, null));
 
         Assert.Equal("sourcePath", exception.ParamName);
-        Assert.Contains("Source path cannot be null or empty", exception.Message);
+        Assert.Contains("FileName path cannot be null", exception.Message);
```
Comment (HasanCPS) [inline backend/CPS.ComplexCases.Egress.Tests/Integration/EgressStorageClientTests.cs:237]: This should be `sourcePath `as that's the method parameter name and what we have in the main branch. This looks like it might be a merge conflict issue that needs to be resolved. Could you double-check this against the main branch and resolve if it's a merge conflict?

File: backend/CPS.ComplexCases.Egress.Tests/Unit/EgressStorageClientTests.cs [hunk 2]
```diff
@@ -84,7 +83,7 @@ public async Task OpenReadStreamAsync_WithValidParameters_ReturnsStream()
         );
 
         // Act
-        await using var result = await _client.OpenReadStreamAsync(path, workspaceId, fileId);
```
Comment (HasanCPS) [inline backend/CPS.ComplexCases.Egress.Tests/Unit/EgressStorageClientTests.cs:87]: same here

File: backend/CPS.ComplexCases.Egress/Client/EgressStorageClient.cs [hunk 3]
```diff
@@ -30,19 +31,17 @@ public async Task<Stream> OpenReadStreamAsync(string path, string? workspaceId,
         var response = await SendRequestAsync(_egressRequestFactory.GetWorkspaceDocumentRequest(arg, token));
         return await response.Content.ReadAsStreamAsync();
     }
+
     public async Task<UploadSession> InitiateUploadAsync(string destinationPath, long fileSize, string? workspaceId = null, string? sourcePath = null, TransferOverwritePolicy? overwritePolicy = null)
     {
         var token = await GetWorkspaceToken();
 
-        if (string.IsNullOrEmpty(sourcePath))
```
Comment (HasanCPS) [inline backend/CPS.ComplexCases.Egress/Client/EgressStorageClient.cs:37]: I believe this is due to a merge conflict. We need to keep this logic here as the line below is checking/using it, and this was suggested and implemented as part of the file transfer PR.

File: backend/CPS.ComplexCases.Egress/Client/EgressStorageClient.cs [hunk 4]
```diff
@@ -104,4 +104,106 @@ public async Task CompleteUploadAsync(UploadSession session, string? md5hash = n
 
         await SendRequestAsync(_egressRequestFactory.CompleteUploadRequest(completeArg, token));
     }
+
+    public async Task<IEnumerable<FileTransferInfo>> ListFilesForTransferAsync(List<TransferEntityDto> selectedEntities, string? workspaceId = null, int? caseId = null)
+    {
+        if (string.IsNullOrEmpty(workspaceId))
```
Comment (HasanCPS) [inline backend/CPS.ComplexCases.Egress/Client/EgressStorageClient.cs:110]: Could you add validation for the `selectedEntities `parameter? Currently, if it's null or empty, we'll get a runtime exception:
```
if (selectedEntities == null || !selectedEntities.Any())
    throw new ArgumentException("Selected entities cannot be null or empty", nameof(selectedEntities));
```

File: backend/CPS.ComplexCases.Egress/Client/EgressStorageClient.cs [hunk 5]
```diff
@@ -104,4 +104,106 @@ public async Task CompleteUploadAsync(UploadSession session, string? md5hash = n
...
+        {
+            throw new ArgumentNullException(nameof(workspaceId), "Workspace ID cannot be null or empty.");
```
Comment (HasanCPS) [inline backend/CPS.ComplexCases.Egress/Client/EgressStorageClient.cs:112]: we're using `ArgumentNullException `but checking for both null and empty. Should this be `ArgumentException `instead?

`throw new ArgumentException("Workspace ID cannot be null or empty.", nameof(workspaceId));`

File: backend/CPS.ComplexCases.Egress/Client/EgressStorageClient.cs [hunk 6]
```diff
@@ -104,4 +104,106 @@ public async Task CompleteUploadAsync(UploadSession session, string? md5hash = n
...
+                GetAllFilesFromFolderParallel(workspaceId, folder.Id, token));
+
+            var subFolderResults = await Task.WhenAll(subFolderTasks);
+            files.AddRange(subFolderResults.SelectMany(x => x));
+        }
+
+        return files;
+    }
+
+    private async Task<List<ListCaseMaterialDataResponse>> GetAllPagesInParallel(string workspaceId, string? folderId, string token)
```
Comment (HasanCPS) [inline backend/CPS.ComplexCases.Egress/Client/EgressStorageClient.cs:167]: The parallel pagination looks efficient! One thought - should we consider limiting concurrent requests to avoid overwhelming the external API, especially for large folder structures?

File: backend/CPS.ComplexCases.Egress/Client/EgressStorageClient.cs [hunk 7]
```diff
@@ -104,4 +104,106 @@ public async Task CompleteUploadAsync(UploadSession session, string? md5hash = n
...
+            {
+                return await GetAllFilesFromFolderParallel(workspaceId, entity.Id, token);
+            }
+        });
+
+        var results = await Task.WhenAll(entityTasks);
+        return results.SelectMany(files => files);
+    }
+
+    private async Task<List<FileTransferInfo>> GetAllFilesFromFolderParallel(string workspaceId, string? folderId, string token)
```
Comment (HasanCPS) [inline backend/CPS.ComplexCases.Egress/Client/EgressStorageClient.cs:140]: The `GetAllFilesFromFolderParallel `method could potentially hit infinite recursion with circular folder references. Would it make sense to add a depth limit as a safety measure?

File: backend/CPS.ComplexCases.FileTransfer.API/Functions/ListFilesForTransfer.cs [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+using CPS.ComplexCases.Common.Models.Domain.Dtos;
+using CPS.ComplexCases.Common.Models.Domain.Enums;
+using CPS.ComplexCases.Common.Models.Requests;
+using CPS.ComplexCases.FileTransfer.API.Factories;
+using CPS.ComplexCases.FileTransfer.API.Models.Domain;
+using CPS.ComplexCases.FileTransfer.API.Validators;
+
+namespace CPS.ComplexCases.FileTransfer.API.Functions;
+
+public class ListFilesForTransfer(ILogger<ListFilesForTransfer> logger, IStorageClientFactory storageClientFactory)
```
Comment (HasanCPS) [inline backend/CPS.ComplexCases.FileTransfer.API/Functions/ListFilesForTransfer.cs:17]: Great work on the implementation! I noticed we don't have unit tests for the `ListFilesForTransfer` function yet. Given the complexity of the logic, adding test coverage would be really valuable.

**Suggested Test Cases:**

**High Priority Tests:**
```csharp
[Fact]
public async Task Run_EgressToNetApp_WithInvalidPaths_SetsIsInvalidCorrectly()
// Tests the validation logic on line 49

[Fact]
public async Task Run_WithValidRequest_ReturnsExpectedResult()
// Happy path test

[Fact]
public async Task Run_WithInvalidRequest_ReturnsBadRequest()
// Tests validation error handling
```

**Additional Coverage:**
```csharp
[Fact]
public async Task Run_NonEgressDirection_SkipsPathValidation()
// Ensures validation only runs for EgressToNetApp

[Fact]
public async Task Run_WithEmptySourcePaths_HandlesGracefully()
// Edge case testing

[Fact]
public async Task Run_MapsTransferEntityDtoCorrectly()
// Tests the DTO mapping logic
```

This will help ensure the function behaves correctly and make future changes safer!

File: backend/CPS.ComplexCases.FileTransfer.API/Validators/FilePathValidator.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+            return false;
+
+        int index = path.LastIndexOf('/');
+        if (index != -1)
+        {
+            var filePath = path[..index]; // Get the file path part before the last '/'
+            var fileName = path[(index + 1)..]; // Get the file name part after the last '/'
+
+            char[] invalidFilePathChars = Path.GetInvalidFileNameChars();
+            char[] invalidFileNameChars = Path.GetInvalidPathChars();
```
Comment (HasanCPS) [inline backend/CPS.ComplexCases.FileTransfer.API/Validators/FilePathValidator.cs:39]: These arrays are created for every validation call. Consider making them static readonly:

```
private static readonly char[] InvalidPathChars = Path.GetInvalidPathChars();
private static readonly char[] InvalidFileNameChars = Path.GetInvalidFileNameChars();
```

File: backend/CPS.ComplexCases.FileTransfer.API/Validators/FilePathValidator.cs [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+                    .NotEmpty()
+                    .WithMessage("File path cannot be empty.")
+                    .MaximumLength(255)
+                    .WithMessage(x => $"{x.Path}: exceeds the 255 characters limit.")
+                    .Must(pathValue => !HasInvalidFileNameChars(pathValue))
+                    .WithMessage(x => $"{x.Path}: contains invalid characters.");
+            });
+    }
+
+    private static bool HasInvalidFileNameChars(string path)
```
Comment (HasanCPS) [inline backend/CPS.ComplexCases.FileTransfer.API/Validators/FilePathValidator.cs:27]: The current `HasInvalidFileNameChars` method has a few issues that this change would address:

**Problems:**
- **Critical Bug**: Character arrays are swapped on lines 38-39
- **Performance**: Arrays recreated on every call, O(n) lookups
- **Cross-platform**: Hard-coded `/` won't work on Windows

**Suggested Solution:**
```csharp
private static readonly HashSet<char> InvalidPathCharsSet = [..Path.GetInvalidPathChars()];
private static readonly HashSet<char> InvalidFileNameCharsSet = [..Path.GetInvalidFileNameChars()];

private static bool HasInvalidPathChars(string path)
{
    if (string.IsNullOrEmpty(path)) return false;

    var directory = Path.GetDirectoryName(path);
    var fileName = Path.GetFileName(path);

    var hasInvalidDirChars = !string.IsNullOrEmpty(directory) && 
                             directory.Any(c => InvalidPathCharsSet.Contains(c));
    var hasInvalidFileChars = fileName.Any(c => InvalidFileNameCharsSet.Contains(c));

    return hasInvalidDirChars || hasInvalidFileChars;
}
```

This fixes the bug, improves performance with HashSet lookups, and uses .NET's cross-platform path methods. Thoughts?

File: backend/Cps.ComplexCases.Common/Extensions/PathExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+public static class PathExtensions
+{
+    public static string RemovePathPrefix(this string path, string? prefix)
+    {
+        if (prefix == null)
+        {
+            return path;
+        }
+        if (path.StartsWith(prefix, StringComparison.CurrentCultureIgnoreCase))
```
Comment (HasanCPS) [inline backend/Cps.ComplexCases.Common/Extensions/PathExtensions.cs:11]: For file paths, consider using `StringComparison.OrdinalIgnoreCase` instead of `CurrentCultureIgnoreCase`. File paths should be compared consistently regardless of user locale:

`if (path.StartsWith(prefix, StringComparison.OrdinalIgnoreCase))`

File: backend/Cps.ComplexCases.Common/Extensions/PathExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+        return path;
+    }
+
+    public static string EnsureTrailingSlash(this string path)
+    {
+        if (string.IsNullOrEmpty(path))
+        {
+            return string.Empty;
+        }
+        return path.EndsWith(Path.DirectorySeparatorChar.ToString(), StringComparison.CurrentCultureIgnoreCase) ? path : path + Path.DirectorySeparatorChar;
```
Comment (HasanCPS) [inline backend/Cps.ComplexCases.Common/Extensions/PathExtensions.cs:24]: Converting char to string for comparison is unnecessary. Also, same culture issue:

`return path.EndsWith(Path.DirectorySeparatorChar) ? path : path + Path.DirectorySeparatorChar;`

File: backend/CPS.ComplexCases.NetApp/Client/NetAppStorageClient.cs [hunk 13]
```diff
@@ -69,4 +72,68 @@ public async Task<UploadChunkResult> UploadChunkAsync(UploadSession session, int
...
+            return null;
+        }
+
+        filesForTransfer.AddRange(
+            response.Data.FileData.Select(x => new FileTransferInfo
+            {
+                SourcePath = x.Path
+            }));
+
+        if (response.Pagination.NextContinuationToken != null)
```
Comment (HasanCPS) [inline backend/CPS.ComplexCases.NetApp/Client/NetAppStorageClient.cs:128]: Recursive calls could cause stack overflow with many pages. Consider iterative approach:

```
string? continuationToken = null;
do
{
    var response = await _netAppClient.ListObjectsInBucketAsync(arg);
    // Process response
    continuationToken = response.Pagination.NextContinuationToken;
} while (continuationToken != null);
```

File: backend/Cps.ComplexCases.Common/Helpers/RequestValidator.cs [hunk 14]
```diff
@@ -1,13 +1,13 @@
...
-using Microsoft.AspNetCore.Http;
 using System.Text.Json;
+using Microsoft.AspNetCore.Http;
 using CPS.ComplexCases.Common.Models;
+using FluentValidation;
 
 namespace CPS.ComplexCases.Common.Helpers;
 
-public static class ValidatorHelper
+public class RequestValidator : IRequestValidator
```
Comment (HasanCPS) [inline backend/Cps.ComplexCases.Common/Helpers/RequestValidator.cs:8]: I see you're using `IRequestValidator `instead of the static `ValidatorHelper`. Is this a refactoring you've done to make the code more testable? If so, that's a great improvement!

Comment (NeilFoubister) [reply backend/Cps.ComplexCases.Common/Helpers/RequestValidator.cs:8]: Yes, that correct! Thanks

File: backend/CPS.ComplexCases.API/Functions/Transfer/InitiateTransfer.cs [hunk 15]
```diff
@@ -1,28 +1,29 @@
...
-public class InitiateTransfer(ILogger<InitiateTransfer> logger, IFileTransferClient transferClient)
+public class InitiateTransfer(ILogger<InitiateTransfer> logger, IFileTransferClient transferClient, IRequestValidator requestValidator)
 {
     private readonly ILogger<InitiateTransfer> _logger = logger;
     private readonly IFileTransferClient _transferClient = transferClient;
+    private readonly IRequestValidator _requestValidator = requestValidator;
 
     [Function(nameof(InitiateTransfer))]
-    [OpenApiOperation(operationId: nameof(Run), tags: ["File Transfer"], Description = "Initiate a file transfer.")]
+    [OpenApiOperation(operationId: nameof(Run), tags: ["FileTransfer"], Description = "Initiate a file transfer.")]
```
Comment (HasanCPS) [inline backend/CPS.ComplexCases.API/Functions/Transfer/InitiateTransfer.cs:26]: Could we also update the tag name for the `GetTransferStatus` function?

File: backend/CPS.ComplexCases.Egress.Tests/Unit/EgressStorageClientTests.cs [hunk 16]
```diff
@@ -20,7 +21,7 @@
 
 namespace CPS.ComplexCases.Egress.Tests.Unit;
 
-public class EgressStorageClientTests : IDisposable
```
[the remark this answers is not in the document]
Comment (HasanCPS) [reply backend/CPS.ComplexCases.Egress.Tests/Unit/EgressStorageClientTests.cs:23]: I think you might have missed that by mistake.
